From: Chad Monroe <chad@monroe.io>
Date: Wed, 30 Oct 2024 19:51:06 -0700
Subject: [PATCH] net: ethernet: mediatek: add PPE roaming workaround  

if neighbor info for a LAN client changes, such as when it moves from one
physical port to another, flush any related PPE entries.	

this will cause the affected flows to revert to software path where the
new neighbor info can be learned.

originally sourced from the Mediatek SDK[1].

1: https://git01.mediatek.com/plugins/gitiles/openwrt/feeds/mtk-openwrt-feeds/+/1023cdd0d0f7d32615474a1262b155cb94f9baac%5E%21/

Signed-off-by: Chad Monroe <chad@monroe.io>
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.c |   12 +
 drivers/net/ethernet/mediatek/mtk_eth_soc.h |    3
 drivers/net/ethernet/mediatek/mtk_ppe.c     |  183 ++++++++++++++++++
 drivers/net/ethernet/mediatek/mtk_ppe.h     |    2
 4 files changed, 199 insertions(+), 1 deletion(-)

--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -3513,7 +3513,6 @@ static void mtk_gdm_config(struct mtk_et
 	mtk_w32(eth, val, MTK_GDMA_FWD_CFG(id));
 }
 
-
 static bool mtk_uses_dsa(struct net_device *dev)
 {
 #if IS_ENABLED(CONFIG_NET_DSA)
@@ -3600,6 +3599,13 @@ static int mtk_open(struct net_device *d
 		for (i = 0; i < ARRAY_SIZE(eth->ppe); i++)
 			mtk_ppe_start(eth->ppe[i]);
 
+		if (eth->soc->offload_version) {
+			err = mtk_ppe_roaming_start(eth);
+			if (err)
+				netdev_err(dev, "%s: could not start ppe roaming work: %d\n",
+						__func__, err);
+		}   
+
 		for (i = 0; i < MTK_MAX_DEVS; i++) {
 			if (!eth->netdev[i])
 				continue;
@@ -3728,6 +3734,10 @@ static int mtk_stop(struct net_device *d
 	for (i = 0; i < ARRAY_SIZE(eth->ppe); i++)
 		mtk_ppe_stop(eth->ppe[i]);
 
+	if (eth->soc->offload_version) {
+		mtk_ppe_roaming_stop(eth);
+	}
+
 	if (mac->pextp)
 		phy_power_off(mac->pextp);
 
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.h
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.h
@@ -1417,6 +1417,9 @@ struct mtk_eth {
 
 	struct mtk_ppe			*ppe[3];
 	struct rhashtable		flow_table;
+	struct socket			*ppe_roam_sock;
+	struct work_struct		ppe_roam_work;
+	unsigned char			ppe_roam_buf[1024];
 
 	struct bpf_prog			__rcu *prog;
 
--- a/drivers/net/ethernet/mediatek/mtk_ppe.c
+++ b/drivers/net/ethernet/mediatek/mtk_ppe.c
@@ -10,6 +10,7 @@
 #include <linux/if_vlan.h>
 #include <net/dst_metadata.h>
 #include <net/dsa.h>
+#include <net/netfilter/nf_flow_table.h>
 #include "mtk_eth_soc.h"
 #include "mtk_ppe.h"
 #include "mtk_ppe_regs.h"
@@ -493,6 +494,29 @@ int mtk_foe_entry_set_queue(struct mtk_e
 	return 0;
 }
 
+static bool
+mtk_foe_mac_match(struct mtk_eth *eth, struct mtk_foe_entry *entry, u8 *mac)
+{
+	int type;
+
+	type = mtk_get_ib1_pkt_type(eth, entry->ib1);
+	if (type > MTK_PPE_PKT_TYPE_IPV4_DSLITE) {
+		if(((swab32(entry->ipv6.l2.dest_mac_hi) == *(u32 *)mac) &&
+		    (swab16(entry->ipv6.l2.dest_mac_lo) == *(u16 *)&mac[4])) ||
+		   ((swab32(entry->ipv6.l2.src_mac_hi) == *(u32 *)mac) &&
+		    (swab16(entry->ipv6.l2.src_mac_lo) == *(u16 *)&mac[4])))
+			return true;
+	} else {
+		if(((swab32(entry->ipv4.l2.dest_mac_hi) == *(u32 *)mac) &&
+		    (swab16(entry->ipv4.l2.dest_mac_lo) == *(u16 *)&mac[4])) ||
+		   ((swab32(entry->ipv4.l2.src_mac_hi) == *(u32 *)mac) &&
+		    (swab16(entry->ipv4.l2.src_mac_lo) == *(u16 *)&mac[4])))
+			return true;
+	}
+
+	return false;
+}
+
 static int
 mtk_flow_entry_match_len(struct mtk_eth *eth, struct mtk_foe_entry *entry)
 {
@@ -1116,3 +1140,162 @@ int mtk_ppe_stop(struct mtk_ppe *ppe)
 
 	return 0;
 }
+
+static int mtk_ppe_delete_by_mac(struct mtk_ppe *ppe, u8 *mac)
+{
+	struct mtk_eth *eth;
+	int i, j, count = 0;
+
+	if (!ppe)
+		return 0;
+
+	eth = ppe->eth;
+
+	for (i = 0; i < MTK_PPE_ENTRIES; i++) {
+		struct mtk_foe_entry *entry = mtk_foe_get_entry(ppe, i);
+		struct flow_offload_tuple tuple;
+		int type, state;
+
+		state = FIELD_GET(MTK_FOE_IB1_STATE, entry->ib1);
+		if (state != MTK_FOE_STATE_BIND || !mtk_foe_mac_match(eth, entry, mac))
+			continue;
+
+		memset(&tuple, 0, sizeof(tuple));
+
+		if (entry->ib1 & MTK_FOE_IB1_UDP)
+			tuple.l4proto = IPPROTO_UDP;
+		else
+			tuple.l4proto = IPPROTO_TCP;
+
+		type = mtk_get_ib1_pkt_type(eth, entry->ib1); 
+		if (type > MTK_PPE_PKT_TYPE_IPV4_DSLITE) {
+			tuple.l3proto = NFPROTO_IPV6;
+			tuple.src_port = htons(entry->ipv6.src_port);
+			tuple.dst_port = htons(entry->ipv6.src_port);
+			for (j = 0; j < 4; j ++) {
+				tuple.src_v6.s6_addr32[j] = htonl(entry->ipv6.src_ip[j]);
+				tuple.dst_v6.s6_addr32[j] = htonl(entry->ipv6.dest_ip[j]);
+			}
+		} else {
+			tuple.l3proto = NFPROTO_IPV4;
+			tuple.src_port = htons(entry->ipv4.orig.src_port);
+			tuple.dst_port = htons(entry->ipv4.orig.dest_port);
+			tuple.src_v4.s_addr = htonl(entry->ipv4.orig.src_ip);
+			tuple.dst_v4.s_addr = htonl(entry->ipv4.orig.dest_ip);
+		}
+
+		flow_offload_teardown_by_tuple(&tuple);
+		count++;
+
+		dev_dbg(eth->dev, "%s: mac=%pM entry=%x deleted from PPE", __func__, mac, i);
+	}
+
+	dev_dbg(eth->dev, "%s: no entry for mac=%pM found in PPE", __func__, mac);
+
+	return count;
+}
+
+static void mtk_ppe_roam_handler(struct work_struct *work)
+{
+	struct mtk_eth *eth = container_of(work, struct mtk_eth, ppe_roam_work);
+	struct kvec iov;
+	struct msghdr msg;
+	struct nlmsghdr *nlh;
+	struct ndmsg *ndm;
+	struct nlattr *nla;
+	int len, ifindex, i;
+	u8 mac[ETH_ALEN];
+
+	if (!eth->ppe_roam_sock)
+		return;
+
+	iov.iov_base = eth->ppe_roam_buf;
+	iov.iov_len = sizeof(eth->ppe_roam_buf);
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_namelen = sizeof(struct sockaddr_nl);
+
+	len = kernel_recvmsg(eth->ppe_roam_sock, &msg, &iov, 1, iov.iov_len, 0);
+	if (len <= 0)
+		goto out;
+
+	nlh = (struct nlmsghdr*)eth->ppe_roam_buf;
+	if (!NLMSG_OK(nlh, len) || nlh->nlmsg_type != RTM_NEWNEIGH)
+		goto out;
+
+	len = nlh->nlmsg_len - NLMSG_HDRLEN;
+	ndm = (struct ndmsg *)NLMSG_DATA(nlh);
+	if (ndm->ndm_family != PF_BRIDGE)
+		goto out;
+
+	ifindex = ndm->ndm_ifindex;
+	nla = (struct nlattr *)((unsigned char *)ndm + sizeof(struct ndmsg));
+	len -= NLMSG_LENGTH(sizeof(struct ndmsg));
+
+	while (nla_ok(nla, len)) {
+		if (nla_type(nla) == NDA_LLADDR) {
+			memcpy(mac, nla_data(nla), ETH_ALEN);
+
+			for (i = 0; i < ARRAY_SIZE(eth->ppe); i++) {
+				if (!eth->ppe[i])
+					continue;
+
+				mtk_ppe_delete_by_mac(eth->ppe[i], mac);
+			}
+
+			dev_dbg(eth->dev, "%s: neighbor=%pM updated", __func__, mac);
+		}
+
+		nla = nla_next(nla, &len);
+	}
+
+out:
+	schedule_work(&eth->ppe_roam_work);
+}
+
+int mtk_ppe_roaming_start(struct mtk_eth *eth)
+{
+	struct socket *sock = NULL;
+	struct sockaddr_nl addr;
+	int ret;
+
+	INIT_WORK(&eth->ppe_roam_work, mtk_ppe_roam_handler);
+
+	ret = sock_create_kern(&init_net, AF_NETLINK, SOCK_RAW, NETLINK_ROUTE, &sock);
+	if (ret < 0)
+		goto out;
+
+	eth->ppe_roam_sock = sock;
+
+	addr.nl_family = AF_NETLINK;
+	addr.nl_pad = 0;
+	addr.nl_pid = 65534;
+	addr.nl_groups = 1 << (RTNLGRP_NEIGH - 1);
+	ret = kernel_bind(sock, (struct sockaddr *)&addr, sizeof(addr));
+	if (ret < 0)
+		goto out;
+
+	schedule_work(&eth->ppe_roam_work);
+
+	dev_info(eth->dev, "mtk_ppe: roaming work activated");
+
+	return 0;
+
+out:
+	if (sock)
+		sock_release(sock);
+
+	return ret;
+}
+
+int mtk_ppe_roaming_stop(struct mtk_eth *eth)
+{
+	if (!eth->ppe_roam_sock)
+		return -ENOENT;
+
+	sock_release(eth->ppe_roam_sock);
+	eth->ppe_roam_sock = NULL;
+
+	dev_info(eth->dev, "mtk_ppe: roaming work deactivated");
+
+	return 0;
+}
--- a/drivers/net/ethernet/mediatek/mtk_ppe.h
+++ b/drivers/net/ethernet/mediatek/mtk_ppe.h
@@ -349,6 +349,8 @@ struct mtk_ppe *mtk_ppe_init(struct mtk_
 void mtk_ppe_deinit(struct mtk_eth *eth);
 void mtk_ppe_start(struct mtk_ppe *ppe);
 int mtk_ppe_stop(struct mtk_ppe *ppe);
+int mtk_ppe_roaming_start(struct mtk_eth *eth);
+int mtk_ppe_roaming_stop(struct mtk_eth *eth);
 int mtk_ppe_prepare_reset(struct mtk_ppe *ppe);
 struct mtk_foe_accounting *mtk_ppe_mib_entry_read(struct mtk_ppe *ppe, u16 index);
 
